[ags4] add string Split and Join

ericoporto opened the pull request:
adds String Split - it's working afaict. I am leaving as a draft because I would like to add String Join but I don't know yet how to read a String array.

ivan-mogilko commented:
> but I don't know yet how to read a String array.

String dynamic array is an array of managed handles; these are 32-bit values, which may be resolved to a String object using `ccGetObjectAddressFromHandle`.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -257,6 +258,55 @@ const char* String_UpperCase(const char *thisString) {
+
+    // this takes a vector of const char*, which means null terminated array of strings
+    // null terminated means we have to hold copies instead of references in-string
+    DynObjectRef arr = DynamicArrayHelpers::CreateStringArray(items);
```
This DynamicArrayHelpers::CreateStringArray was originally implemented for strings retrieved from somewhere else, but since you already create ScriptString::Buffers here, you can add another variant, which accepts `std::vector<ScriptString> &&` and moves this data to ScriptString::Create().

ivan-mogilko replied:
Another alternative would be a variant of DynamicArrayHelpers::CreateArray that accepts a generic std::vector<DynObjectRef> &`. In this case you could create ScriptStrings here directly using ScriptString::Create, and then pass DynObjectRefs into CreateArray, where it makes array, and stores handles in there.

I cannot say which is more convenient at the moment.

ericoporto replied:
> since you already create ScriptString::Buffers here, you can add another variant,

I can't figure how to iterate the buffer array from this because I keep getting the following error:

> Call to implicitly-deleted copy constructor of 'ScriptString::Buffer' copy constructor is implicitly deleted because 'Buffer' has a user-declared move constructor

> you could create ScriptStrings here directly using ScriptString::Create, and then pass DynObjectRefs into CreateArray

This worked, so I went this way.

ericoporto replied:
Oh so it was double ampersand in the call like `&&items`, that's why it didn't work. :O

ericoporto replied:
Sorry, decided to fix that "FIX-ME" and I tried the following:

```
    String type_str = !items.empty() ? items[0].Mgr->GetType() : "";
    auto type = std::find_if(ccInstance::GetRTTI()->GetTypes().begin(), ccInstance::GetRTTI()->GetTypes().end(),
                             [type_str](const RTTI::Type& i) -> bool {return type_str.Equals(i.name); });
```

This works and `type` correctly finds the type "String" in the "_BuiltInScriptHeader.ash", but the type size is 0 - instead of `sizeof(int32_t)`, so I can't use `CreateNew` because I don't understand how to correctly get the size dynamically.

ivan-mogilko replied:
The thing is, RTTI may be missing. It has to be decided, whether this API actually requires type info or not.

ivan-mogilko replied:
For faster searching for a type there's a lookup table - ccInstance::_rttiLookup, but it may be not exposed right now. May be worth adding an inlined function that returns a constant reference to this table.

ericoporto replied:
It's alright, it is working with the CreateOld method, I can update that in a new PR. But thanks for the information!

The find wasn't as slow. About "fully qualified", this means just "_BuiltInScriptHeader.ash::String" or something that can be determined like this right? If so, then it should be fine anyway, at least for internal usage.

ivan-mogilko replied:
About this, type's size is 0 because there are no plain fields in it. RTTI only knows the fields declared in script, and there are none in String.

But when you make dynamic array of pointers, you don't need type's size at all, as you don't store type itself in the array, you store pointers to that type. So you ever need only size of pointer. If you were making a dynamic array of primitive types or plain (non-managed) structs, then you would have to get their size.

ericoporto replied:
Right, I think this means I need to check the ARRAY_MANAGED_TYPE_FLAG in this case. This flag is found in the type_id here in ags4 and was found in the elementCount in ags3 I think.

I think I understand now how to do it but I think I will leave this for later.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -257,6 +258,55 @@ const char* String_UpperCase(const char *thisString) {
+        std::copy(ptr, found_cstr, buf.Get());
+        buf.Get()[len] = 0;
+        buflist.push_back(std::move(buf));
+        items.push_back(buflist.back().Get());
```
This code not make much sense, `ScriptString::CreateBuffer` is meant for creating a specialized memory buffer holding scriptstring's meta-data. The purpose is to have a buffer that may be directly moved to ScriptString::Create. If you want only copies of strings, then create simple Strings.

Although here you'd rather make a std::vector<ScriptString::Buffer> and move it further, see another comment below.

ivan-mogilko commented:
Looking at the draft code, I think you misunderstand the purpose of ScriptString::Buffer. This Buffer object is meant for the case when you need to fill it in multiple steps, copying parts of other strings over, as opposed to a single string assignment.

Second thing is, this Buffer is for passing directly into ScriptString::Create, avoiding extra copies and length recalculation. Having it as a temporary string storage but not moved further defeats its purpose. For that you could using simple Strings (and they are easier to assign to, no need to use std::copy).

---

But here, for your purpose, it would make more sense to either
1. Write another variant of `DynamicArrayHelpers::CreateStringArray` that accepts `std::vector<ScriptString::Buffer> &` and std::moves its members to `ScriptString::Create` calls.
2. Write another variant of `DynamicArrayHelpers::CreateArray` that accepts `std::vector<DynObjectRef> &`; create ScriptStrings here directly using ScriptString::Create, and then pass DynObjectRefs into CreateArray, where it makes array, and stores handles in there.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -257,6 +258,55 @@ const char* String_UpperCase(const char *thisString) {
+        len = found_cstr - ptr;
+        auto buf = ScriptString::CreateBuffer(len, 0);
+        std::copy(ptr, found_cstr, buf.Get());
+        buf.Get()[len] = 0;
```
I think this could be simplified by using a single `snprintf(buf.Get(), "%s", ptr, len + 1);`

This also makes me wonder why I haven't added an assignment operator to ScriptString::Buffer, which would just copy an input string over...

ericoporto commented:
> String dynamic array is an array of managed handles; these are 32-bit values, which may be resolved to a String object using `ccGetObjectAddressFromHandle`.

Thanks, I think I got it, I will try to update in the weekend with Join.

ericoporto commented:
I got String.Join to work!!!

> String dynamic array is an array of managed handles; these are 32-bit values, which may be resolved to a String object using `ccGetObjectAddressFromHandle`.

Thanks for this because without it, I wouldn't get anywhere! I will leave as a draft still because I want to see if I can improve it's code still.

ericoporto commented:
I got code that works and I think the code in general is looking ok now - but any possible changes please tell me!!

Other things to review!

- Do we want this? I feel these are useful extensions, but it's good to check how everyone feels.
- is the API ok? Please check the names of things in the API, parameter order, ...
- @AlanDrake suggested adding a flag for Trim in String.Split too, but maybe this would complicate the code significantly - it would need to be trimmed before allocating the buffer but using a specific function that did not copy the buffer but instead selected things... I liked the idea at first but now I would prefer this to be done in AGS Script as a String extension
  - I feel some additional string stuff could be in a script library of sorts...

ivan-mogilko commented:
> @AlanDrake suggested adding a flag for Trim in String.Split too, but maybe this would complicate the code significantly - it would need to be trimmed before allocating the buffer but using a specific function that did not copy the buffer but instead selected things...

I don't think it would complicate much. You are already using 2 pointers: for beginning of a string and end of string. Trimming the source would mean moving these pointers right and left correspondingly.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -49,6 +49,67 @@ int String_IsNullOrEmpty(const char *thisString)
     return 0;
 }
 
+void calculate_joint_length(const size_t count, const uint32_t *&handles, int sep_len, int sep_ulen, int &total_len, int &total_ulen)
```
I notice that you iterate over handles and do `ccGetObjectAddressFromHandle` on them twice in succession, first to calculate final length and second to actually do string copying. This seem suboptimal and complicates the code. You could instead make a function that gathers the resolved ScriptString's pointers into a std::vector, and then do whatever you want with them: get their headers, check lengths, then copy their contents.

ericoporto replied:
I got rid of the function but wasn't sure on making it a function because then I would iterate three times on the strings, because I still need to figure the length before I can create a buffer...

ivan-mogilko replied:
Hmm... well, the function itself would iterate over handles and call ccGetObjectAddressFromHandle in order to gather strings, that may come handy later. Other 2 loops would be over a ready vector of string pointers, I don't think that's too bad.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -49,6 +49,67 @@ int String_IsNullOrEmpty(const char *thisString)
+        if (address != nullptr) {
+            auto const& sh = ScriptString::GetHeader(address);
+            memcpy(p, address, sh.Length);
+            p += sh.Length;
```
if you use std::copy, it returns iterator after copied fragment, letting you write slightly less lines:
```p = std::copy(address, address +sh.Length, p);```

ericoporto replied:
good, I didn't knew this, changed but didn't push force - will squash commits things later

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -49,6 +49,67 @@ int String_IsNullOrEmpty(const char *thisString)
+            total_len += sh.Length;
+            total_ulen += sh.ULength;
+        }
+        if (i != count - 1) {
```
I think you may simplify this by not adding separator lengths in a loop, but add them once afterwards, since this only depends on the `count`:
```
total_len += sep_len * (count - 1);
total_ulen += sep_ulen * (count - 1);
```

ericoporto replied:
that's true, I didn't notice that when writing the code.

ivan-mogilko commented:
There's a curious question about null arguments in Join. I noticed that passing null array would result in returning null pointer, but having null Strings in that array have same effect as empty strings. Which means that if I pass an array of only nulls, that would return an actual string, which may be empty or containing a sequence of separators.

Just a observation...

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -257,6 +258,52 @@ const char* String_UpperCase(const char *thisString) {
+    if ((separator == nullptr) || (separator[0] == 0)) {
+        std::vector<const char*> items{};
+        items.push_back(thisString);
+        DynObjectRef arr = DynamicArrayHelpers::CreateStringArray(std::vector<const char*>(items));
```
This is probably a typo, should be passing `items`, because right now you are constructing a new temporary vector.

ericoporto replied:
thanks! This was leftover from a code that did all in one line - but looked ugly...

ivan-mogilko commented:
Well, I find mostly minor things that may be optimized. I will be testing this properly tomorrow, but I guess it should be working.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -49,6 +49,62 @@ int String_IsNullOrEmpty(const char *thisString)
+    int sep_len, sep_ulen;
+    ustrlen2(separator, &sep_len, &sep_ulen);
+
+    std::vector<ScriptString*> script_strings{};
```
This is not correct, the ScriptString is only a manager, but the registered objects are buffers of type `char*`.

ericoporto replied:
Ah, this is why the autocomplete was making no sense with the data.

ivan-mogilko commented:
I need to honestly say that I will want to pick out an algorithm of resolving handles to object pointers as a separate function, that may be used in any API that has a dynamic array argument. I believe that's better to have this algorithm standalone rather than to mix it with data processing.

ericoporto commented:
If you write this function and put it somewhere in ags4 I can rebase and adjust the code to use it.

ivan-mogilko commented:
> If you write this function and put it somewhere in ags4 I can rebase and adjust the code to use it.

Okay, I wrote couple of variants, for cases when you do and do not need to know actual "managers": 387aa5e8ed8844e1e379f9dd4fb63749fb97284c

ericoporto commented:
@ivan-mogilko  I tried to adjust in the commit https://github.com/adventuregamestudio/ags/pull/2401/commits/99d408038b6912c2219e11f17c5a86e6ff988834

> EDIT: in regards to this PR, I tested that String.Join correctly recreates a string after String.Split, and it works as expected.

This exact case is failing for me in my commit, the `assert(header.IsPointerArray());` is failing, so I think I did something wrong building the string. :/

ivan-mogilko commented:
I just tried your latest commit, and it still works for me. It looks 99% like my own test (the only difference is that some typecast is not needed).

Can you post the script you're using for a test?

ericoporto commented:
ah, crap, that is exactly it, I recompiled the game but did not place in the right folder...

ivan-mogilko commented on Engine/ac/dynobj/cc_dynamicarray.cpp:
```diff
@@ -189,6 +189,28 @@ DynObjectRef DynamicArrayHelpers::CreateStringArray(std::vector<ScriptString::Bu
+    auto type = std::find_if(ccInstance::GetRTTI()->GetTypes().begin(), ccInstance::GetRTTI()->GetTypes().end(),
+                             [type_str](const RTTI::Type& i) -> bool {return type_str.Equals(i.name); });
+
+    uint32_t size = type->size == 0 ? sizeof(int32_t) : type->size;
```
This is a wrong size calculation. If you are creating array of pointers, that's always int32_t, regardless of whether type has size >0 or not. On another hand, size == 0 does not indicate a pointer storage.

ericoporto replied:
I adjusted it here https://github.com/adventuregamestudio/ags/pull/2401/commits/82488f296cb4e62b12933d7a82615fde1e0220d4 , I think dynamically in the api we will always be creating a dynamic array of pointers?

ivan-mogilko commented on Engine/ac/dynobj/cc_dynamicarray.h:
```diff
@@ -93,7 +93,9 @@ namespace DynamicArrayHelpers
     DynObjectRef CreateStringArray(const std::vector<const char*>);
     // Create array of managed strings from script strings
-    DynObjectRef CreateStringArray(std::vector<ScriptString::Buffer> &&);
+    DynObjectRef CreateStringArrayFromBuffers(std::vector<ScriptString::Buffer> &&);
```
There's really no need to have different name, it was fine as a function overload.

ericoporto replied:
I had to remove the overload because the linter I use was not picking up the overload kept telling me this function was unused, I wasn't comfortable with it.

ivan-mogilko commented:
Please note, that you may also use CCDynamicArray::CreateOld if you don't want to bother with type names.
For array of pointers you simply can pass "is_managed" as true.

ivan-mogilko commented:
> fix crash if the string goes out of scope but not the array

I think I understand what the problem was.
You were not doing AddRef on the source string when you place it into array.

ivan-mogilko commented:
Could you elaborate the meaning of "fix scope crashes" commit?

Comment sais:
> fix crash if the string goes out of scope but not the array

Do you refer to C++ code, or AGS script?

ericoporto commented:
Sorry, I am readjusting things, but the problem is if you write in AGS Script something like this:
```AGS Script
  String splitted[];
  {
    String str = "text,to,split";
    splitted = str.Split(",");
  }
  for(int i=0; i<splitted.Length; i++)
  {
    System.Log(eLogInfo, "'%s'",  splitted[i]);
  }
```

ivan-mogilko commented:
The PR looks fine overall now. There's just one thing: since you added CreateArray helper, you could just use that instead of CreateStringArrayFromBuffers, if you'd create ScriptStrings from these buffers right away. Anyway, there's a room of improvement in the existing helper functions, and string factory methods.

ericoporto commented:
There's no rush to this and it doesn't block anything, so it's ok to take time. There was also no previous issue discussing it.

ivan-mogilko commented:
Well, I tried few script tests, and things appear working, including returning arrays from functions, splitting by multiple char sequences, etc. I may merge this now, and leave any possible code simplification for the later.
